File: backend/app/schemas/review.py
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, Field


class ReviewSeverity(str, Enum):
    high = "high"
    medium = "medium"
    low = "low"


class ReviewCategory(str, Enum):
    logic = "logic"
    style = "style"
    grammar = "grammar"
    citation = "citation"
    structure = "structure"
    hallucination = "hallucination"
    figure = "figure"
    code = "code"
    consistency = "consistency"
    other = "other"


class ReviewActionType(str, Enum):
    replace = "replace"
    delete = "delete"
    insert = "insert"
    revise = "revise"


class ReviewIssueStatus(str, Enum):
    open = "open"
    applied = "applied"
    resolved = "resolved"
    dismissed = "dismissed"
    stale = "stale"
    failed = "failed"


class ReviewDimensionOut(BaseModel):
    key: str | None = None
    dimension: str | None = None
    label: str = ""
    weight: int = 0
    raw_score: int = Field(default=0, ge=0, le=100)
    effective_score: int = Field(default=0, ge=0, le=100)
    score: int = Field(default=0, ge=0, le=100)
    issue_count: int = 0
    summary: str = ""
    detector: str = ""
    confidence: float = Field(default=0.0, ge=0, le=1)
    status: str = "completed"


class ReviewIssueOut(BaseModel):
    id: str
    severity: ReviewSeverity
    category: ReviewCategory = ReviewCategory.other
    title: str
    detail: str = ""
    quote: str = ""
    suggestion: str = ""
    action_type: ReviewActionType = ReviewActionType.replace
    paragraph_index: int | None = None
    char_offset: int | None = None
    dimension: str | None = None
    issue_type: str = "review_issue"
    penalty: int = 0
    status: ReviewIssueStatus = ReviewIssueStatus.open
    explanation: str = ""
    action: ReviewActionType = ReviewActionType.replace
    replacement_text: str = ""
    paragraph_id: str | None = None
    char_start: int | None = None
    char_end: int | None = None
    anchor_hash: str | None = None
    issue_fingerprint: str | None = None
    quality_evidence: dict[str, Any] | None = None
    detector: str = "review_agent"
    confidence: float = Field(default=0.0, ge=0, le=1)
    stale: bool = False


class ReviewApplyIssueIn(BaseModel):
    action_type: ReviewActionType
    quote: str = Field(default="", max_length=2000)
    suggestion: str = Field(default="", max_length=4000)
    detail: str = Field(default="", max_length=4000)
    context: str = Field(default="", max_length=12000)


class ReviewIssuePreviewOut(BaseModel):
    issue_id: str | None = None
    application_id: str | None = None
    quote: str
    result_text: str
    result_markdown: str | None = None
    preview_kind: Literal["replace", "insert", "delete"]
    diff: dict[str, Any] = {}
    locator_strategy: str = ""
    locator_confidence: float = 0.0
    preview_required: bool = False
    stale: bool = False
    affected_dimensions: list[str] = []
    score_changes: list[dict[str, Any]] = []
    warning: dict[str, Any] | None = None
    paragraph_id: str | None = None
    paragraph_index: int | None = None
    char_start: int | None = None
    char_end: int | None = None


class ReviewApplyIssueOut(ReviewIssuePreviewOut):
    pass


class ReviewConfirmApplicationOut(BaseModel):
    application_id: str
    issue_status: ReviewIssueStatus | None = None
    score: int | None = None
    dimensions: list[ReviewDimensionOut] = []


class ChapterReviewIn(BaseModel):
    """可选传入正文；缺省则使用数据库中该章已保存内容。"""

    text: str | None = Field(default=None, max_length=120_000)


class CitationLintIssueOut(BaseModel):
    kind: str
    quote: str = ""
    detail: str = ""
    suggested_title: str | None = None


class ChapterReviewOut(BaseModel):
    chapter_index: int
    chapter_title: str
    summary: str
    score: int = Field(ge=0, le=100)
    total_score: int | None = Field(default=None, ge=0, le=100)
    dimensions: dict[str, ReviewDimensionOut] = {}
    dimension_rows: list[ReviewDimensionOut] = []
    issues: list[ReviewIssueOut]
    citation_issues: list[CitationLintIssueOut] = []
    word_count: int = 0
    review_id: str | None = None
    snapshot_hash: str | None = None
    snapshot_md: str | None = None
    status: str = "completed"
    is_stale: bool = False
    created_at: str | None = None


class ReviewHistoryItemOut(BaseModel):
    review_id: str
    chapter_index: int
    chapter_title: str
    score: int
    status: str
    snapshot_hash: str
    created_at: str
    is_stale: bool = False
    dimensions: list[ReviewDimensionOut] = []


class IssueStatusOut(BaseModel):
    issue: ReviewIssueOut
    review: ChapterReviewOut | None = None


class AiInlineSelectionIn(BaseModel):
    from_: int | None = Field(default=None, alias="from")
    to: int | None = None
    text: str = Field(..., min_length=1, max_length=500)
    paragraph_id: str | None = None


class AiInlinePreviewIn(BaseModel):
    selection: AiInlineSelectionIn
    instruction: str = Field(default="使表达更清晰、自然。", max_length=2000)
    context_before: str = Field(default="", max_length=4000)
    context_after: str = Field(default="", max_length=4000)


class AiInlinePreviewOut(BaseModel):
    preview_id: str
    original_text: str
    rewritten_text: str
    diff: dict[str, Any] = {}
    validation: dict[str, Any] = {}

File: frontend/src/types/review.ts
export type ReviewSeverity = "high" | "medium" | "low";
export type ReviewActionType = "replace" | "delete" | "insert" | "revise";
export type ReviewIssueStatus = "open" | "applied" | "resolved" | "dismissed" | "stale" | "failed";

export type ReviewCategory =
  | "logic"
  | "style"
  | "grammar"
  | "citation"
  | "structure"
  | "hallucination"
  | "figure"
  | "code"
  | "consistency"
  | "other";

export type ReviewDimensionKey =
  | "logic_structure"
  | "language_grammar"
  | "style_consistency"
  | "citation_sources"
  | "factual_support"
  | "figure_quality"
  | "ai_signature";

export interface ReviewIssue {
  id: string;
  severity: ReviewSeverity;
  category: ReviewCategory;
  title: string;
  detail: string;
  quote: string;
  suggestion: string;
  action_type?: ReviewActionType;
  paragraph_index?: number | null;
  char_offset?: number | null;
  dimension?: ReviewDimensionKey | string | null;
  issue_type?: string;
  penalty?: number;
  status?: ReviewIssueStatus;
  explanation?: string;
  action?: ReviewActionType;
  replacement_text?: string;
  paragraph_id?: string | null;
  char_start?: number | null;
  char_end?: number | null;
  anchor_hash?: string | null;
  issue_fingerprint?: string | null;
  detector?: string;
  confidence?: number;
  stale?: boolean;
}

export interface ReviewDimension {
  key?: ReviewDimensionKey | string | null;
  dimension?: ReviewDimensionKey | string | null;
  label?: string;
  weight?: number;
  raw_score?: number;
  effective_score?: number;
  score: number;
  issue_count?: number;
  summary?: string;
  detector?: string;
  confidence?: number;
  status?: string;
}

export interface ReviewApplyResult {
  issue_id?: string | null;
  application_id?: string | null;
  quote: string;
  result_text: string;
  result_markdown?: string | null;
  preview_kind: "replace" | "insert" | "delete";
  diff?: Record<string, unknown>;
  locator_strategy?: string;
  locator_confidence?: number;
  preview_required?: boolean;
  stale?: boolean;
  affected_dimensions?: string[];
  score_changes?: Array<Record<string, unknown>>;
  warning?: Record<string, unknown> | null;
  paragraph_id?: string | null;
  paragraph_index?: number | null;
  char_start?: number | null;
  char_end?: number | null;
}

export interface ReviewConfirmResult {
  application_id: string;
  issue_status?: ReviewIssueStatus | null;
  score?: number | null;
  dimensions?: ReviewDimension[];
}

export interface CitationLintIssue {
  kind: string;
  quote: string;
  detail: string;
  suggested_title?: string | null;
}

export interface ChapterReviewResult {
  chapter_index: number;
  chapter_title: string;
  summary: string;
  score: number;
  total_score?: number | null;
  dimensions?: Record<string, ReviewDimension>;
  dimension_rows?: ReviewDimension[];
  issues: ReviewIssue[];
  citation_issues?: CitationLintIssue[];
  word_count: number;
  review_id?: string | null;
  snapshot_hash?: string | null;
  snapshot_md?: string | null;
  status?: string;
  is_stale?: boolean;
  created_at?: string | null;
}

export interface ReviewHistoryItem {
  review_id: string;
  chapter_index: number;
  chapter_title: string;
  score: number;
  status: string;
  snapshot_hash: string;
  created_at: string;
  is_stale?: boolean;
  dimensions?: ReviewDimension[];
}

export interface AiInlinePreviewResult {
  preview_id: string;
  original_text: string;
  rewritten_text: string;
  diff?: Record<string, unknown>;
  validation?: Record<string, unknown>;
}

export const REVIEW_DIMENSION_LABEL: Record<ReviewDimensionKey, string> = {
  logic_structure: "逻辑结构",
  language_grammar: "语言语法",
  style_consistency: "风格一致",
  citation_sources: "引用来源",
  factual_support: "事实支撑",
  figure_quality: "图表质量",
  ai_signature: "AI味风险",
};

export const REVIEW_CATEGORY_LABEL: Record<ReviewCategory, string> = {
  logic: "逻辑",
  style: "文风",
  grammar: "语病",
  citation: "引用",
  structure: "结构",
  hallucination: "事实支撑",
  figure: "图表",
  code: "代码",
  consistency: "风格一致",
  other: "其他",
};

export const REVIEW_ACTION_LABEL: Record<ReviewActionType, string> = {
  replace: "替换",
  delete: "删除",
  insert: "新增",
  revise: "AI 改写",
};

export const REVIEW_SEVERITY_LABEL: Record<ReviewSeverity, string> = {
  high: "严重",
  medium: "中等",
  low: "轻微",
};

export const REVIEW_STATUS_LABEL: Record<ReviewIssueStatus, string> = {
  open: "待处理",
  applied: "已应用",
  resolved: "已解决",
  dismissed: "已忽略",
  stale: "旧版本",
  failed: "失败",
};
